### PR TITLE
Group cycle data together and use PreImageCache

### PR DESCRIPTION
**Pull request by Geod24**
See each commit individually. There are still a few things to fix (e.g. `nonce` is not used) but I'll leave it for a later PR.

### CODE REVIEW

File: source/agora/consensus/EnrollmentManager.d [hunk 1]
```diff
@@ -278,6 +287,8 @@ public class EnrollmentManager
         this.enroll_key = frozen_utxo_hash;
 
         // N, cycle length
+        // Note, this name iis misleading, as a cycle here is the length
```
Comment (AndrejMitrovic) [inline source/agora/consensus/EnrollmentManager.d]:
```suggestion
        // Note, this name is misleading, as a cycle here is the length
```

File: source/agora/consensus/EnrollmentManager.d [hunk 2]
```diff
@@ -118,8 +118,17 @@ public class EnrollmentManager
...
+
+        The number of enrollments we can emit in a cycle
+
+        With the current setting of ValidatorCycle being 1008 and blocks
+        having 10 minutes of interval, this gives us roughly 700 days,
+        or two years, per cycle.
+
+    ***************************************************************************/
+
+    private immutable uint EnrollPerCycle = 100;
```
Comment (AndrejMitrovic) [inline source/agora/consensus/EnrollmentManager.d]: I'm not sure this name is that much better. The name seems to suggest there will be 100 enrollments per cycle, and a cycle is 1008 blocks? But that doesn't make sense.

File: source/agora/consensus/EnrollmentManager.d [hunk 3]
```diff
@@ -287,12 +330,12 @@ public class EnrollmentManager
         this.enroll_key = frozen_utxo_hash;
 
         // N, cycle length
-        // Note, this name iis misleading, as a cycle here is the length
+        // Note, this name is misleading, as a cycle here is the length
```
Comment (AndrejMitrovic) [inline source/agora/consensus/EnrollmentManager.d]: It looks like you introduced the spelling mistake in the previous commit, and fixed it in this commit.

File: source/agora/consensus/EnrollmentManager.d (hunk 2 above)
Comment (Geod24) [reply source/agora/consensus/EnrollmentManager.d]: A cycle is 100 enrollments, and enrollment is 1008 blocks.

Comment (linked0) [reply source/agora/consensus/EnrollmentManager.d]: We use a cycle in white paper as the period where a node is validator. So how about `CyclesPerNonce`?

**Comment by Geod24** [comment]
Removed the first commit, and the typo by extension

**Review by linked0 — APPROVED** [review]
LGTM.
I checked the code throughly. It's very structural change.

**Review by AndrejMitrovic — APPROVED** [review]

**Pull request merged**